server/src/main/java/org/opensearch/search/aggregations/bucket/composite/CompositeAggregator.java

```diff
@@ -153,12 +164,49 @@ final class CompositeAggregator extends BucketsAggregator {
         }
         this.queue = new CompositeValuesCollectorQueue(context.bigArrays(), sources, size, rawAfterKey);
         this.rawAfterKey = rawAfterKey;
+
+        // Try fast filter optimization when the only source is date histogram
+        if (sourceConfigs.length == 1 && sourceConfigs[0].valuesSource() instanceof RoundingValuesSource) {
+            RoundingValuesSource dateHistogramSource = (RoundingValuesSource) sourceConfigs[0].valuesSource();
+            bucketOrds = LongKeyedBucketOrds.build(context.bigArrays(), CardinalityUpperBound.ONE);
+            preparedRounding = dateHistogramSource.getPreparedRounding();
+            long afterValue = 0;
+            if (rawAfterKey != null) {
+                assert rawAfterKey.size() == 1 && formats.size() == 1;
+                afterValue = formats.get(0).parseLong(rawAfterKey.get(0).toString(), false, () -> {
+                    throw new IllegalArgumentException("now() is not supported in [after] key");
+                });
```

Comment:
The logic is very date histogram specific. Can we abstract this out better to not pollute generic `CompositeAggregator` class?

Reply:
Refactor out to the Helper